pkg/sql/binlog_subscriber.go

```diff
@@ -0,0 +1,616 @@
+package sql
+
+import (
+	"context"
+	"database/sql"
+	"encoding/binary"
+	"encoding/json"
+	"fmt"
+	"math/rand"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/pkg/errors"
+	"github.com/siddontang/go-mysql/canal"
+	"github.com/siddontang/go-mysql/mysql"
+
+	"github.com/ThreeDotsLabs/watermill"
+	"github.com/ThreeDotsLabs/watermill/message"
+)
+
+var (
+	ErrBinlogSubscriberClosed = errors.New("binlog subscriber is closed")
+)
+
+// TODO: probably this should be a part of schema adapter, it is a temporary shortcut
+type ColumnsMapping struct {
+	UUID     string
+	Offset   string
+	Payload  string
+	Metadata string
+}
+
+type Database struct {
+	Host     string
+	Port     string
+	User     string
+	Password string
+	Name     string
+	Flavour  string
+}
+
+type BinlogSubscriberConfig struct {
+	ConsumerGroup string
+
+	// ResendInterval is the time to wait before resending a nacked message.
+	// Must be non-negative. Defaults to 1s.
+	ResendInterval time.Duration
+
+	// SchemaAdapter provides the schema-dependent queries and arguments for them, based on topic/message etc.
+	SchemaAdapter ExtendedSchemaAdapter
+
+	// OffsetsAdapter provides mechanism for saving acks and offsets of consumers.
+	OffsetsAdapter BinlogOffsetAdapter
+
+	// InitializeSchema option enables initializing schema on making subscription.
+	InitializeSchema bool
+
+	// ColumnsMapping provides mechanism for mapping columns in tables to message.
+	ColumnsMapping ColumnsMapping
+
+	// Database provides mechanism for passing credentials to the database.
+	Database Database
+}
+
+func (c *BinlogSubscriberConfig) setDefaults() {
+	if c.ResendInterval == 0 {
+		c.ResendInterval = time.Second
+	}
+	if c.Database.Host == "" {
+		c.Database.Host = "localhost"
+	}
+	if c.Database.Port == "" {
+		c.Database.Port = "3306"
+	}
+	if c.Database.User == "" {
+		c.Database.User = "root"
+	}
+	if c.Database.Name == "" {
+		c.Database.Name = "watermill"
+	}
+	if c.Database.Flavour == "" {
+		c.Database.Flavour = mysql.MySQLFlavor
+	}
+	if c.ColumnsMapping.UUID == "" {
+		c.ColumnsMapping.UUID = "uuid"
+	}
+	if c.ColumnsMapping.Payload == "" {
+		c.ColumnsMapping.Payload = "payload"
+	}
+	if c.ColumnsMapping.Metadata == "" {
+		c.ColumnsMapping.Metadata = "metadata"
+	}
+	if c.ColumnsMapping.Offset == "" {
+		c.ColumnsMapping.Offset = "offset"
+	}
+}
+
+func (c BinlogSubscriberConfig) validate() error {
+	if c.ResendInterval <= 0 {
+		return errors.New("resend interval must be a positive duration")
+	}
+	if c.SchemaAdapter == nil {
+		return errors.New("schema adapter is nil")
+	}
+	if c.OffsetsAdapter == nil {
+		return errors.New("offsets adapter is nil")
+	}
+	if c.Database.Flavour != mysql.MySQLFlavor && c.Database.Flavour != mysql.MariaDBFlavor {
+		return errors.New("database config incorrect flavour value")
```

Comment:
The error message could list the allowed flavours.
Otherwise, the consumer of the error will have to arrive to this line in the source code to find out what values are allowed.